Restore dropped sentence in description of "discriminate" (#19623)
coq/coq: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

   in the form :n:`@term__1 <> @term__2` that are inconsistent
   (:ref:`example <discriminate_goal_inequality_ex>`).

>    If :n:`@induction_arg` provides a :n:`@one_term_with_bindings`, only that
>    term is checked for inconsistency.

Reviewer: ```suggestion
   If :n:`@induction_arg` is provided, only the provided proof term
   or hypothesis is checked for inconsistency.
```

Author: > Indeed, Jason is right.

Oh??  What is the proper terminology for the output of the `Show Proof` command?  I've been using "proof term" to mean something that satisfies `type of (proof term) = some term`.  Hypotheses don't have proof terms in this sense; they are merely types.  Given that `discriminate` is looking for contradictions in equalities such as, say, 1=2, it's odd to suggest that a hypothesis is a proven statement.  It would be rather bad to have a proof (proof term) for the hypothesis `False`.

If the `induction_arg` is a more complex expression, I expect `discriminate` is evaluating it but is not computing its type.

@JasonGross > "Suppose we have a proof that x = y; call that proof H."

This makes a supposition for the context of a paper.  I don't see how it's really relevant to `discriminate`.

``
Goal 1=1.
reflexivity.
Show Proof.   (*  eq_refl *)
```

Reviewer 2: > Hypotheses don't have proof terms in this sense; they are merely types.

This is wrong.  Hypotheses *have* types, in exactly the same way that terms have types. 

> @JasonGross > "Suppose we have a proof that x = y; call that proof H."
> 
> This makes a supposition for the context of a paper.

This is exactly what `intro H` does when the goal is `x = y -> ...`.  The supposition is a hypothesis named `H` which has type `x = y`.

Reviewer 2: > Given that `discriminate` is looking for contradictions in equalities such as, say, 1=2, it's odd to suggest that a hypothesis is a proven statement

We might say: "Let us show that 1 = 2 is not provable.  We will do this by assuming we have a proof of `1 = 2`, call it `H`, and showing that from this presumption we can prove `False`.  Define the function `f : nat -> Prop` so that `f x` is `True` is `x` is 1 and `False` otherwise.  We can trivially prove `f 1`, call this proof `F`.  But by (rewriting with) `H` (in the type of `F`), we get a proof of `f 2`, that is, `False`.  Qed.". This proof is roughly what is constructed when you do `intro H; discriminate` on the goal `1 = 2 -> False`.

> It would be rather bad to have a proof (proof term) for the hypothesis `False`.

Only in the empty context.  In a contradictory context, it's fine to have a proof term of type `False` (and this is exactly what discriminate creates from the contradictory context)

Author: Let's take this step by step and see if we can agree on some things.

>  "Suppose we have a proof that x = y; call that proof H."

Why wouldn't you say "Suppose that x = y; call that H."?

> Hypotheses have types, in exactly the same way that terms have types.

What's the useful distinction between a hypothesis having a type and being a type?

```
Goal 1=2 -> True.
intro.
let i := type of H in idtac i.   (* 1=2 - useful for discriminate *)
let i := type of (1=2) in idtac i.   (* Prop - not useful for discriminate *)
```

Would you respond to my comment above?  How do you define "proof term"?

> What is the proper terminology for the output of the `Show Proof` command?  I've been using "proof term" to mean something that satisfies `type of (proof term) = some term`.

Reviewer 2: > Why wouldn't you say "Suppose that x = y; call that H."?

What I wrote is more precise, and more constructive.  Assuming the second "that" refers to the supposition, I'd say that proofs are mathematical objects, but suppositions are not.  

> What's the useful distinction between a hypothesis having a type and being a type?

The difference between having a type and being a type is fundamental to Coq and dependently typed proof assistants.  If you don't understand this, please ask on https://proofassistants.stackexchange.com/, or Zulip, or search on Google, or see if another part of the Refman explains it.  If you're just asking about how the distinction is useful in the case of hypotheses: if we are not precise with our words, then sentences have ambiguous meanings, and I think this makes them more confusing.  There's a difference between "the type T" and "the type of T" (the first says "T, which is by the way a type" while the second says "type of T").  If you're asking for code that gets at the distinction, I'd say
```coq
Goal 1=2 -> forall P : Prop, P.
intros H P.
let i := H in idtac i.  (* H, the hypothesis *)
let i := type of H in idtac i.   (* 1=2, the type of the hypothesis *)
let i := P in idtac i.  (* P, the hypothesis *)
let i := type of P in idtac i.   (* Prop, the type of the hypothesis *)
```

Reviewer 2: > How do you define "proof term"?

I don't think it has a precise technical meaning.  I'd say "an object whose dual Curry-Howard nature as both a proof and a term (program) is being emphasized".

> What is the proper terminology for the output of the `Show Proof` command?

"proof term" seems fine

> I've been using "proof term" to mean something that satisfies `type of (proof term) = some term`.

What is "some term" here?  It is always the case that `type of x = y` for some `y` whenever `x` is a valid (well-typed) term.

Reviewer: I didn't read the exchange in details. Please just let me know when this PR is ready to merge and if you are taking this suggestion of improvement into account. (I think its thread was used to discuss something else.)

Author: >> I've been using "proof term" to mean something that satisfies `type of (proof term) = some term`.

> What is "some term" here?  It is always the case that `type of x = y` for some `y` whenever `x` is a valid (well-typed) term.

Poor choice of words.  I should have said that the proof term PT of a term T satisfies `type of (PT) = T`.  Also that PT is a proof of T.

Reviewer 2: > > > I've been using "proof term" to mean something that satisfies `type of (proof term) = some term`.
> 
> > What is "some term" here?  It is always the case that `type of x = y` for some `y` whenever `x` is a valid (well-typed) term.
> 
> Poor choice of words. I should have said that the proof term PT of a term T satisfies `type of (PT) = T`. Also that PT is a proof of T.

Yes, this usage sounds right to me

Reviewer 2: And note that for a hypothesis named `H`, `H` is a term (whose kernel representation is `Var "H"`, cf https://github.com/coq/coq/blob/1d77a8830d03d3c050af9b9442e707b8cfcf2576/kernel/constr.mli#L244-L247 https://github.com/coq/coq/blob/1d77a8830d03d3c050af9b9442e707b8cfcf2576/user-contrib/Ltac2/Constr.v#L56)

Reviewer 2: Indeed, sorry for hijacking this thread @Zimmi48 .  Your suggestion here seems good to me, and I don't think @jfehrle objects either.

Author: I was confused by @herbelin's [statement](https://github.com/coq/coq/pull/19623#issuecomment-2385294983) "The term `(Z.lt_trans _ _ _ Hn (Pos2Z.neg_is_neg p)` is, after reduction, a proof of `Gt = Lt` (presumably), which is discriminable."  I took "reduction" as a reference to [that term in the doc](https://coq.inria.fr/doc/V8.20.0/refman/language/core/conversion.html#term-reduction), which is about conversion rules and I skipped over the "a proof of".

But it must be reductions using the typing rules:

`let i := type of (Z.lt_trans _ _ _ Hn (Pos2Z.neg_is_neg p)) in idtac i.    (*   (0 < 0)%Z   *)`

Some tactics such as `apply` can take a proof term because they compute the type.  IIUC `rewrite` does not.  But I've never tried to review all the tactics and figure out which are which and that their descriptions are correct and covered systematically.  Probably most are in the `apply` camp?  I don't recall seeing a good general description of this.  Also, the manual doesn't really explain how to construct proof terms on your own.  It seems like an infrequently used feature.  How important is it to document rolling your own proof terms?

Author: @Zimmi48 OK, this is ready.  Would you respond to the paragraph just above "Some tactics ..."?

The fact the tactic computes the type of complex expressions while (in effect) it doesn't do that for a hypothesis `H` should explained somewhere in the doc (not here).

Reviewer 2: > Some tactics such as `apply` can take a proof term because they compute the type. IIUC `rewrite` does not.

`rewrite` can take proof terms, you can do `rewrite (eq_trans ...)`

> Also, the manual doesn't really explain how to construct proof terms on your own. 

I believe the "core language" section explains how to construct terms, and there's no (technical) difference between "term" and "proof term".

> But it must be reductions using the typing rules:
> 
> `let i := type of (Z.lt_trans _ _ _ Hn (Pos2Z.neg_is_neg p)) in idtac i. (* (0 < 0)%Z *)`

Yes, consider
```coq
let i := type of (Z.lt_trans _ _ _ Hn (Pos2Z.neg_is_neg p)) in
let v := (eval cbv in i) in
idtac v.
```

> The fact the tactic computes the type of complex expressions while (in effect) it doesn't do that for a hypothesis `H` should be explained somewhere in the doc (not here).

Certainly there is something to be explained, but I'm not sure it is this.  The algorithm does not distinguish between hypotheses and complex expressions (except perhaps for parsing), and doesn't need to.  That it appears to be doing something distinct for terms is a confusion that should probably be resolved somewhere, just by noting that hypotheses are a special case of terms.  (There are some operations that can be done with hypotheses that cannot be done with general terms, such as renaming them, clearing them, rewriting *in* them, etc.  But any operation that can be done with terms can also be done with hypotheses.)

Reviewer: > It seems like an infrequently used feature. How important is it to document rolling your own proof terms?

In Agda, there are no tactics. All proofs are built as proof terms. In Lean, they support switching between building proof terms manually and using tactics more easily, but I think they mostly use tactics. In Coq, we are mostly using tactics at the large scale. But at the small scale, it is frequent to build proof terms, e.g., when applying a hypothesis to specific arguments before using it, etc. It is an important feature.
